cpp/CMakeLists.txt

```diff
@@ -837,3 +810,10 @@ config_summary_message()
 if(${ARROW_BUILD_CONFIG_SUMMARY_JSON})
   config_summary_json()
 endif()
+
+# Write out arrow and third party package configurations.
+
+configure_file("src/arrow/config.h.cmake" "src/arrow/config.h")
```

Comment:
1. You don't want to pollute your source tree with this, otherwise it'll break when one is using multiple build configuration (I do), see [ARROW_ADD_PKG_CONFIG](https://github.com/apache/arrow/blob/master/cpp/cmake_modules/BuildUtils.cmake#L803-L807) for something similar. This works for the `util/config.h` because there's nothing of importance in there. But it'll definitively break with define guards, e.g. ARROW_WITH_DATASET.
2. `.cmake` is not idiomatic, use `.in`, every package manager / OSS / C developer will understand. You can replace the one in `util` too.

Reply:
1. This directive doesn't put the generated file into your source tree, it goes to a location in the cmake build directory which overrides "src/arrow/config.h"
2. The file includes `#cmakedefine` directives, which do the reasonable thing with an `ON/OFF` value. I can refactor to use `.in` if you prefer

Reply:
I'll retract my point on `.cmake`, just leave it to minimize the diff.